torch/csrc/jit/tensorexpr/kernel.cpp

```diff
@@ -1554,6 +1563,64 @@ Tensor* TensorExprKernel::computeSum(const torch::jit::Value* v) {
       reduction_info.reductionDims);
 }
 
+Tensor* TensorExprKernel::computeSoftmax(const torch::jit::Value* v) {
+  // Softmax is computed as follows:
+  //    softmax(vi) = exp(vi) / sum(exp(vi))
+  //
+  // This is implemented as 3 loopnests:
+  //   - First loop computes exp for every element in v.
+  //   - Second loop computes the sum over the softmax dim.
+  //   - Final loop computes softmax for every element in v.
+
+  TORCH_INTERNAL_ASSERT(v->node()->inputs().size() == 3);
+  auto output_dims = dimsFromSizes(sizesForValue(v));
+
+  // We do not handle None for dims (input 1) because that is supposed to
+  // be deprecated.
+  TORCH_INTERNAL_ASSERT(v->node()->input(1)->node()->kind() == prim::Constant);
+  size_t softmax_dim = v->node()->input(1)->node()->i(attr::value);
+
+  // The sum needed for Softmax should be calculated along the softmax
+  // dimension, which will be dimension to reduce.
+  std::vector<DimArg> dim_args;
+  for (size_t i = 0; i < output_dims.size(); ++i) {
+    if (i != softmax_dim) {
+      dim_args.push_back(output_dims[i]);
+    }
+  }
+
+  auto e = computeOneOperand(
+      "aten_softmax_exp", v, [](const ExprHandle& a) { return exp(a); });
+  auto sum = Reduce(
+      "aten_softmax_sum",
+      dim_args,
+      Sum(),
+      [&](ParameterList& indices) {
+        // Reduce will have the softmax dimension as the inner most loop.
+        // So, the innermost index in the incoming indices will refer to the
+        // softmax dimension. Update the incoming indices by moving that
+        // softmax dimension index to the appropriate position.
+        auto new_indices = indices;
+        for (size_t i = softmax_dim; i < indices.size() - 1; ++i) {
+          new_indices[i + 1] = indices[i];
+        }
+        new_indices[softmax_dim] = indices[indices.size() - 1];
+        return e->call(new_indices);
+      },
+      {output_dims[softmax_dim]});
+  return Compute("aten_softmax", output_dims, [&](ParameterList& indices) {
+    // The sum tensor uses all indices other than the softmax dimension
+    // index.
+    std::vector<ExprHandle> new_indices;
+    for (size_t i = 0; i < indices.size(); ++i) {
+      if (i != softmax_dim) {
+        new_indices.push_back(indices[i]);
+      }
+    }
+    return e->call(indices) / sum->call(new_indices);
```

Comment:
In softmax, both the sum and max need to be computed. All elements should be shifted with the max first. Otherwise, if all your elements are quite large, it quickly leads to overflow. 

The tricky part is that for performant execution, the max and sum need to computed together.

Reply:
Why do we need to compute max here? IIUC, softmax only requires the sum over exp of all the elements along the given dimension. Am I missing something?

I understand that the sum could lead to overflow quickly. But that is a known issue with softmax, right? Isn't that one of the reasons we have `log_softmax`?

Reply:
@navahgar 
> Why do we need to compute max here? IIUC, softmax only requires the sum over exp of all the elements along the given dimension. Am I missing something?

Yes, the definition of softmax contains only the sum of exp of all the elements along the given dimensions.
However, subtracting the max of the elements before applying exponent gives a numerically stable algorithm while the output will match with the softmax definition.

> I understand that the sum could lead to overflow quickly. But that is a known issue with softmax, right? Isn't that one of the reasons we have `log_softmax`?

Right, it is a known issue and it has a know solution as described above. Also the computation of log_softmax uses the max-substraction-trick.

@zheng-xq 
When implementing stable and performant softmax, you should follow the torch softmax implementation:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/SoftMax.cpp#L28-L62

Reply:
Thanks for the explanation @pearu.

Reply:
Added the max-subtraction-trick to the implementation.